Modify rule S3519: Expand and adjust for LaYC (#3160)
SonarSource/rspec: rules/S3519/cfamily/rule.adoc

These overreaching accesses cause some of the most damaging and hard to track defects.
Not only do these faulty accesses constitute undefined behavior, but they frequently introduce security vulnerabilities and open up the program to buffer overflow attacks, too.

== What is the potential impact?

Accessing memory that goes beyond the boundary of the allocated array or buffer results in undefined behavior.

For programs that exercise undefined behavior, the compiler is no longer bound by the language specification.
The application may crash or, even worse, the application may appear to execute correctly while losing data or producing incorrect results.

Out-of-bounds memory accesses cause some of the most damaging and hard-to-track defects.
Unbounded memory accesses may also introduce severe security vulnerabilities.
In case a user of the program can control the contents of a buffer, for instance, a malicious attacker can the buffer overflow to overwrite critical data, execute arbitrary code, or gain unauthorized access to the system.

== How to fix it

> Ensure that memory access are within the bounds of the allocated array or buffer.

Reviewer: Shouldn't thins be plural?: memory accesses
Have you checked the text with Grammarly?

If {cpp} is available, use the `std::array` or `std::vector` types to manage stack or heap-based arrays, respectively and use `std::string` to manage string buffers.

=== Code examples

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
#include <stdio.h>

int main() {
  char buffer[64] = {0};
  if (fgets(buffer, 2048, stdin) != NULL) { // Noncompliant: target buffer may overflow
    printf("buffer contents: %s", buffer);
  }
  return 0;
}
----

==== Compliant solution

[source,cpp,diff-id=1,diff-type=compliant]
----
#include <stdio.h>

int main() {
  char buffer[64] = {0};
  if (fgets(buffer, 64, stdin) != NULL) { // Compliant: reads at most 64 - 1 characters
    printf("buffer contents: %s", buffer);
  }
  return 0;

> }

Author: Yes, I've changed the code snippets in question to the C programming language.

  strncpy(buf1, str, strlen(str));
  memcpy(buf2, buf1, 17); // Noncompliant: buf2 will overflow.
  printf("buf1: %s\n", buf1);
  printf("buf2: %s\n", buf2);
  free(buf1);
  free(buf2);
}
----

==== Compliant solution

[source,cpp,diff-id=2,diff-type=compliant]
----
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void foo() {
  char *buf1 = (char *)malloc(17);
  char *buf2 = (char *)malloc(8);
  memset(buf1, 0, 17);
  memset(buf2, 0, 8);
  const char *const str = "AAAA"
                          "BBBB"
                          "CCCC"
                          "DDDD";
  strncpy(buf1, str, strlen(str));
  // Compliant: copy 7 bytes and leave the 8th terminating null byte ('\0') untouched
  // such that `buf2` can be correctly printed in the subsequent lines.
  memcpy(buf2, buf1, 7);
  printf("buf1: %s\n", buf1);
  printf("buf2: %s\n", buf2);
  free(buf1);
  free(buf2);
}
----

==== Noncompliant code example

[source,cpp,diff-id=3,diff-type=noncompliant]
----
#include <stdio.h>

void foo() {
  char txt[16] = "Hello, World!";
  for (size_t i = 0; i <= sizeof(txt); ++i) { // Noncompliant: index should be between 0 and 15
    printf("%c\n", txt[i]);
  }

> }

Reviewer: Unfortunately, due to loop unrolling (4), we don't detect this: https://godbolt.org/z/ezMs5jhcc
I'd recommend reworking this example.

It's also not a good practice to hardcode the length of the string at the array declaration.

Author: I'll give it a shot and rework it a bit.

----

==== Compliant solution

[source,cpp,diff-id=3,diff-type=compliant]
----
void foo() {
  char txt[16] = "Hello, World!";
  for (size_t i = 0; i < sizeof(txt); ++i) { // Compliant: index is between 0 and 15
    printf("%c\n", txt[i]);
  }
}
----

> === Going the extra mile

Reviewer: Before the extra mile, I think it worths mentioning that one needs to be careful for zero terminating buffers whenever needed as this can be a problem (pitfall) in this context.

Author: Let me add a `Pitfalls` section here.

Reviewer: Excellent addition! I love it.

Buffer overflows occur when a program writes data beyond the boundaries of a buffer and can lead to memory corruption and potential security vulnerabilities.

> To mitigate this risk, developers must carefully manage array and buffer sizes (, use secure coding practices, and employ techniques like input validation and bounds checking).

Reviewer: I think we should improve the chunking of this sentence by splitting it. This way we can express the same without using parens and colons.

In {cpp}, manual array or string, i.e., buffer manipulations are considered a code smell.

> Instead, the `std::array` type should be used to manage stack-based arrays, and the `std::vector` type should be used if a heap-based array is desired.

Reviewer: Do we use the term "stack-based" anywhere else?
I think the accurate term is "automatic storage duration", but I guess yours is still more accessible for the reader.

Author: Yes, I think "stack-based" is easier to swallow here. Especially, since we talk about "heap-based" arrays in the next half of the sentence.

> Besides always carrying their respective sizes, i.e., number of elements, `std::array` and `std::vector` implement many useful member functions such as `begin()` and `end()`, allowing one to safely and conveniently process them using algorithms from the {cpp}'s `<algorithm>` header, for instance.

Reviewer: ```suggestion
Besides always carrying their respective sizes, i.e., the number of elements, `std::array` and `std::vector` implement many useful member functions such as `begin()` and `end()`, allowing one to safely and conveniently process them using algorithms from the {cpp}'s `<algorithm>` header, for instance.
```

Author: This should be fine and more of a flow without the "the" as "i.e." refers to "sizes".

An example is shown in the following:

[source,cpp]
----
#include <algorithm>
#include <array>
#include <iostream>
#include <numeric>
#include <vector>

void bar() {
  // stack-based array
  std::array<int, 8> stack_buf;
  std::fill(stack_buf.begin(), stack_buf.end(), 42);
  for (auto i : stack_buf) {
    std::cout << i << ' ';
  }
  std::cout << '\n';
  std::cout << "sum of stack_buf's values: "
            << std::accumulate(stack_buf.begin(), stack_buf.end(), 0) << '\n';
  // heap-based array
  std::vector<int> heap_buf = {1, 2, 3, 4};
  heap_buf.resize(10);
  std::iota(heap_buf.begin(), heap_buf.end(), 1);
  std::cout << "sum of heap_buf's values: "
            << std::accumulate(heap_buf.begin(), heap_buf.end(), 0) << '\n';
}
----

The `std::string` type should be used to manage buffers, since it guarantees safe buffer manipulations.
Instead of manually concatenating two buffers using `strncat`, for instance, `std::string` allows this operation to be performed in a much more convenient manner as shown in the following code:

[source,cpp]
----
#include <iostream>
#include <string>

> void buz(std::string const &s) {

Reviewer: global: I think the example names should reflect their content.

Author: I agree, but I guess that is off the table as this is not consistently handled in the rules that have been rewritten so far, unfortunately. From what I've seen I believe it's quite the contrary :-(

  std::string t = "Hello, " + s;
  std::cout << t << '\n';
}
----

In addition, the `std::format` function allows one to format strings according to a user-specified format and returns the result as a string as shown in what follows:

[source, cpp]
----
#include <format>
#include <iostream>
#include <string>

void tar(std::string const &s) {
  std::string t = std::format("Hello, World! Greetings {}\n", s);
  std::cout << t << '\n';
}
----

== Resources

=== Conference presentations

> * CppCon 2018 - https://www.youtube.com/watch?v=0S0QgQd75Sw&ab_channel=CppCon[Software Vulnerabilities in C and C++]

Reviewer: This link appears to be dead to me.

Author: Oh, thanks for pointing that out. It was alive a few days ago. YouTube is strange sometimes 😄

Author: That's crazy. The link is alive. It's only dead if you try to access the video using the Island browser. When using ordinary Chrome, it's all fine.

Reviewer 2: Perhaps something to report to IT?

=== Standards

* CERT - https://wiki.sei.cmu.edu/confluence/x/wtYxBQ[ARR30-C. Do not form or use out-of-bounds pointers or array subscripts]
* CERT - https://wiki.sei.cmu.edu/confluence/x/i3w-BQ[STR50-CPP. Guarantee that storage for strings has sufficient space for character data and the null terminator]
* CWE - https://cwe.mitre.org/data/definitions/119[119 Improper Restriction of Operations within the Bounds of a Memory Buffer]
* CWE - https://cwe.mitre.org/data/definitions/121.html[121 Stack-based Buffer Overflow]
* CWE - https://cwe.mitre.org/data/definitions/122.html[122 Heap-based Buffer Overflow]
* CWE - https://cwe.mitre.org/data/definitions/131[131 Incorrect Calculation of Buffer Size]

> * CWE - https://cwe.mitre.org/data/definitions/788[788 Access of Memory Location After End of Buffer]

Reviewer: In some entries you have the `.html` suffix, for others don't.
Please harmonize them.

Memory accesses should be bounded to prevent buffer overflows.

== Why is this an issue?

Array overruns and buffer overflows occur when a memory access accidentally goes beyond the boundary of the allocated array or buffer.

These overreaching accesses cause some of the most damaging and hard to track defects.

> Not only do these faulty accesses constitute undefined behavior, but they frequently introduce security vulnerabilities and open up the program to buffer overflow attacks, too.

Reviewer 2: This sounds like tautology
```suggestion
Not only do these faulty accesses constitute undefined behavior, but they frequently introduce security vulnerabilities, too.
```

== What is the potential impact?

Accessing memory that goes beyond the boundary of the allocated array or buffer results in undefined behavior.

For programs that exercise undefined behavior, the compiler is no longer bound by the language specification.
The application may crash or, even worse, the application may appear to execute correctly while losing data or producing incorrect results.

Out-of-bounds memory accesses cause some of the most damaging and hard-to-track defects.
Unbounded memory accesses may also introduce severe security vulnerabilities.

> In case a user of the program can control the contents of a buffer, for instance, a malicious attacker can the buffer overflow to overwrite critical data, execute arbitrary code, or gain unauthorized access to the system.

Reviewer 2: ```suggestion
In case a user of the program can control the contents of a buffer, for instance, a malicious attacker can use the buffer overflow to overwrite critical data, leak secret or private information, execute arbitrary code, or gain unauthorized access to the system.
```

  char buffer[64] = {0};
  if (fgets(buffer, 64, stdin) != NULL) { // Compliant: reads at most 64 - 1 characters
    printf("buffer contents: %s", buffer);
    return 0;
  }
  return 1;
}
----

==== Noncompliant code example

[source,c,diff-id=2,diff-type=noncompliant]
----
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void foo(void) {
  char *buf1 = (char *)malloc(17);
  char *buf2 = (char *)malloc(8);
  memset(buf1, 0, 17);
  memset(buf2, 0, 8);
  const char *const str = "AAAA"
                          "BBBB"
                          "CCCC"
                          "DDDD";
  strncpy(buf1, str, strlen(str));
  memcpy(buf2, buf1, 17); // Noncompliant: buf2 will overflow.
  printf("buf1: %s\n", buf1);
  printf("buf2: %s\n", buf2);
  free(buf1);
  free(buf2);
}
----

==== Compliant solution

[source,c,diff-id=2,diff-type=compliant]
----
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void foo(void) {
  char *buf1 = (char *)malloc(17);
  char *buf2 = (char *)malloc(8);
  memset(buf1, 0, 17);
  memset(buf2, 0, 8);
  const char *const str = "AAAA"
                          "BBBB"
                          "CCCC"
                          "DDDD";

>   strncpy(buf1, str, strlen(str));

Reviewer 2: This is also far from best practice: if I accidentally add some characters to `str`, I'll overrun `buf1`. I think you should change both non-compliant and compliant examples to allocate `buf1` based on `strlen(str)`

Author: I addressed that issue.

  // Compliant: copy 7 bytes and leave the 8th terminating null byte ('\0') untouched
  // such that `buf2` can be correctly printed in the subsequent lines.
  memcpy(buf2, buf1, 7);
  printf("buf1: %s\n", buf1);
  printf("buf2: %s\n", buf2);
  free(buf1);
  free(buf2);
}
----

==== Noncompliant code example

[source,c,diff-id=3,diff-type=noncompliant]
----
#include <stdio.h>

void bar(void) {
  int array[8] = {1, 2, 3, 4, 5, 6, 7, 8};
  array[8] = 42; // Noncompliant: index should be between 0 and 7

>   for (size_t i = 0; i < 8; ++i) { // Compliant: index is between 0 and 7
>     printf("%d\n", array[i]);

Reviewer 2: Since this is a "**non**compliant" example, why not take the opportunity?
```suggestion
  for (size_t i = 0; i <= 8; ++i) { // Noncompliant: index is between 0 and 7
    printf("%d\n", array[i]); // Noncompliant: index reaches 8 - beyond the last element of the array
```

Reviewer: Because we cannot detect that case. It would require us to loop more than 4 times to reach the OOB access.
I asked for this refactoring [here](https://github.com/SonarSource/rspec/pull/3160#discussion_r1338471315).

  }
}
----

==== Compliant solution

[source,c,diff-id=3,diff-type=compliant]
----
#include <stdio.h>

void bar(void) {
  int array[8] = {1, 2, 3, 4, 5, 6, 7, 8};
  array[7] = 42; // Compliant: index is between 0 and 7

>   for (size_t i = 0; i < 8; ++i) { // Compliant: index is between 0 and 7
>     printf("%d\n", array[i]);

Reviewer 2: ```suggestion
  for (size_t i = 0; i < 8; ++i) {
    printf("%d\n", array[i]); // Compliant: index is between 0 and 7
```

  }
}
----

=== Pitfalls

When managing and manipulating string buffers one needs to take great care that the buffers are correctly terminated with a null byte (`'\0'`).
Failing to correctly null terminate string buffers almost always causes buffer overruns and introduces undefined behavior.

The following faulty program aims at creating a string copy.

> However, the heap-allocated memory buffer for holding the copy is too small since `strlen` calculates the length of its input string but exclude the "invisible" terminating null byte ('\0').

Reviewer 2: ```suggestion
However, the heap-allocated memory buffer for holding the copy is too small since `strlen` calculates the length of its input string but exclude the "invisible" terminating null byte (`'\0'`).
```

The call to `strncpy` also uses the string length computed by `strlen` and the string copy hence lacks a null terminator.
This causes out-of-bounds reads and introduces undefined behavior in the subsequent processing steps.
In this example, the call to `printf` will eventually trigger such erroneous behavior.

> [source,c]

Reviewer 2: I would use the diff-view here too, even though it is outside of "how to fix it", I think it can still be displayed in some products, and in the worst case, diff won't be highlighted

----
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *make_str_copy(const char *const src) {
  size_t src_len = strlen(src);
  char *dst = (char *)malloc(src_len); // buffer too small to hold null terminator

>   if (dst == NULL) {
>     perror("malloc failed");
>     exit(1);
>   }

Reviewer 2: ```suggestion
```
It is not always a good idea to check the `malloc` return value for null...

In any case, it is not the focus of this rule, and you already omitted it from other examples, so I would omit it here too, for brevity

Author: I'm usually in favor of brevity, too. However, I feel the need to correctly handle a "failing" `malloc` in this example. Since both, this rule and our examples target low-level C code, we should highlight that many things can go sideways pretty quickly due to the very basic nature of APIs and functionalities provided by the C standard library---making them rather inconvenient to use at times. Unhandled allocation failures are often overlooked, but when manually handling buffers in C you need to get everything exactly right. The `make_str_copy` function should depict a function that is more on the realistic side.

There is only one other example that uses `malloc` which I would adjust to also have it handled correctly, if you don't mind.

Reviewer 2: I think it might not be a good idea because it is rare that you can do anything useful when your system is out of memory, but these frequent error checks still clutter the code. My reasoning was similar to [this SQ question](https://stackoverflow.com/questions/2248995/is-there-a-need-to-check-for-null-after-allocating-memory-when-kernel-uses-over), and [this answer](https://stackoverflow.com/a/49993528). I do not insist here, as the only downside is the code readability, and this rule deals with the safety so probably should err on the safe side

Instead, the `std::array` type should be used to manage stack-based arrays, and the `std::vector` type should be used if a heap-based array is desired.
Besides always carrying their respective sizes, i.e., number of elements, `std::array` and `std::vector` implement many useful member functions such as `begin()` and `end()`, allowing one to safely and conveniently process them using algorithms from the {cpp}'s `<algorithm>` header, for instance.
An example is shown in the following:

[source,cpp]
----
#include <algorithm>
#include <array>
#include <iostream>
#include <numeric>
#include <vector>

void bar() {
  // stack-based array
  std::array<int, 8> stack_buf;
  std::fill(stack_buf.begin(), stack_buf.end(), 42);
  for (auto i : stack_buf) {
    std::cout << i << ' ';
  }
  std::cout << '\n';
  std::cout << "sum of stack_buf's values: "
            << std::accumulate(stack_buf.begin(), stack_buf.end(), 0) << '\n';
  // heap-based array
  std::vector<int> heap_buf = {1, 2, 3, 4};
  heap_buf.resize(10);
  std::iota(heap_buf.begin(), heap_buf.end(), 1);
  std::cout << "sum of heap_buf's values: "
            << std::accumulate(heap_buf.begin(), heap_buf.end(), 0) << '\n';
}
----

> The `std::string` type should be used to manage buffers, since it guarantees safe buffer manipulations.

Reviewer 2: ```suggestion
The `std::string` type should be used to manage buffers since it facilitates safe buffer manipulations.
```
you can still access string out-of-bounds

Instead of manually concatenating two buffers using `strncat`, for instance, `std::string` allows this operation to be performed in a much more convenient manner as shown in the following code:

[source,cpp]
----
#include <iostream>
#include <string>

void buz(std::string const &s) {
  std::string t = "Hello, " + s;
  std::cout << t << '\n';
}
----

In addition, the `std::format` function allows one to format strings according to a user-specified format and returns the result as a string as shown in what follows:

[source, cpp]
----
#include <format>
#include <iostream>
#include <string>

void tar(std::string const &s) {
  std::string t = std::format("Hello, World! Greetings {}\n", s);
  std::cout << t << '\n';
}
----

> == Resources

Reviewer 2: S5945 - for the discussion about replacing arrays with C++ wrappers
Maybe also S1851 that deprecates pointers

Reviewer 2: Also S5782 - overflow in POSIX functions. BTW, I think some of your examples actually belong to S5782 (e.g., overrun in `printf`)

Author: I've added them to the "Related rules" section.

[source,c,diff-id=6,diff-type=noncompliant]
----
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *make_str_copy(const char *const src) {
  size_t src_len = strlen(src);
  // `src_len + 1` since `strlen` _excludes_ the terminating null byte
  char *dst = (char *)malloc(src_len + 1);
  if (dst == NULL) {
    perror("malloc failed");
    exit(1);
  }
  dst[src_len] = '\0'; // explicitly add a null terminator
  strncpy(dst, src, src_len);
  return dst;
}

int main(void) {
  const char str[] = "Hello, World!";
  char *str_copy = make_str_copy(str);
  printf("%s\n", str_copy); // Compliant: `str_copy` has been correctly null terminated
  // More code that processes `str_copy` ...
  free(str_copy);
  return 0;
}
----

=== Going the extra mile

Buffer overflows occur when a program writes data beyond the boundaries of a buffer and can lead to memory corruption and potential security vulnerabilities.
To mitigate this risk, developers must carefully manage array and buffer sizes.
This includes using secure coding practices, and employing techniques like input validation and bounds checking.

> Various modern C and C++ compilers allow one to automatically instrument a program during compilation using so-called _sanitizers_.

Reviewer: ```suggestion
Various modern C and {cpp} compilers allow one to automatically instrument a program during compilation using so-called _sanitizers_.
```

}
----

The `std::string` type should be used to manage buffers since it facilitates safe buffer manipulations.
Instead of manually concatenating two buffers using `strncat`, for instance, `std::string` allows this operation to be performed in a much more convenient manner as shown in the following code:

[source,cpp]
----
#include <iostream>
#include <string>

void buz(std::string const &s) {
  std::string t = "Hello, " + s;
  std::cout << t << '\n';
}
----

In addition, the `std::format` function allows one to format strings according to a user-specified format and returns the result as a string as shown in what follows:

[source, cpp]
----
#include <format>
#include <iostream>
#include <string>

void tar(std::string const &s) {
  std::string t = std::format("Hello, World! Greetings {}\n", s);
  std::cout << t << '\n';
}
----

== Resources

=== Articles & blog posts

* Clang Hardening - https://blog.quarkslab.com/clang-hardening-cheat-sheet.html[Clang Hardening Cheat Sheet]
* FORTIFY_SOURCE - https://www.redhat.com/en/blog/enhance-application-security-fortifysource[Enhance application security with FORTIFY_SOURCE]

> * Stack Protection - https://developers.redhat.com/articles/2022/06/02/use-compiler-flags-stack-protection-gcc-and-clang#:~:text=Stack%20canary&text=A%20modification%20indicates%20the%20detection,%2Dfstack%2Dprotector%2Dstrong[Use compiler flags for stack protection in GCC and Clang]

Reviewer: ```suggestion
* Stack Protection - https://developers.redhat.com/articles/2022/06/02/use-compiler-flags-stack-protection-gcc-and-clang[Use compiler flags for stack protection in GCC and Clang]
```

Stuff after the `?` or `#` in URLs are usually optional.
